packages/descendants/src/useDescendant.tsx

```diff
@@ -0,0 +1,80 @@
+import {
+  ComponentProps,
+  ForwardedRef,
+  RefObject,
+  useContext,
+  useMemo,
+  useRef,
+} from 'react';
+
+import {
+  useForwardedRef,
+  useIsomorphicLayoutEffect,
+} from '@leafygreen-ui/hooks';
+
+import { DescendantContextType } from './DescendantsContext';
+import { findDescendantIndexWithId } from './utils';
+
+/** @deprecated */
+const genId = () => `_${Math.random().toString(36).substr(2, 9)}`;
+
+interface UseDescendantReturnObject<T extends HTMLElement> {
+  ref: RefObject<T>;
+  index: number;
+  id: string;
+}
+
+/**
+ * Registers a component as a descendant of a given context, and calculates its index within the rendered descendant list.
+ */
+export const useDescendant = <T extends HTMLElement>(
+  context: DescendantContextType<T>,
+  fwdRef?: RefObject<T> | ForwardedRef<T>,
+  props?: ComponentProps<any>,
+): UseDescendantReturnObject<T> => {
+  const ref: React.RefObject<T> = useForwardedRef(fwdRef ?? null, null);
+  const { descendants, dispatch } = useContext(context);
+  const id = useRef(genId());
+
+  // Find the element with this id in the descendants list
+  const index = useMemo(() => {
+    return findDescendantIndexWithId(descendants, id.current);
+  }, [descendants]);
+
+  // On render, register the element as a descendant
+  useIsomorphicLayoutEffect(() => {
```

Comment:
if dispatch or ref changes do we need to change the type of dispatch to 'update'?

Reply:
(i.e. should we do a lookup before creating?)

Reply:
When wither of those props change (which they likely won't, but still), the cleanup function will be run, which calls `"remove"`, and then we re-register it

Reply:
We also do [a check within the reducer](https://github.com/mongodb/leafygreen-ui/blob/a26750588462d711f81222269708dd24e8da4251/packages/descendants/src/DescendantsReducer.ts#L65) to make sure we're not duplicating